WalletWasabi.Fluent/Behaviors/NavBarSelectedIndicatorState.cs

```diff
@@ -0,0 +1,171 @@
+using System.Threading;
+using Avalonia;
+using Avalonia.Animation;
+using Avalonia.Animation.Easings;
+using Avalonia.Controls.Shapes;
+using Avalonia.Layout;
+using Avalonia.Media;
+using Avalonia.Styling;
+using Avalonia.VisualTree;
+
+namespace WalletWasabi.Fluent.Behaviors
+{
+	public class NavBarSelectedIndicatorState : IDisposable
+	{
+		private readonly Easing _bckEasing = new SplineEasing(0.2, 1, 0.1, 0.9);
+		private readonly Easing _fwdEasing = new SplineEasing(0.1, 0.9, 0.2);
+		private readonly TimeSpan _totalDuration = TimeSpan.FromSeconds(0.4);
+
+		private CancellationTokenSource _currentAnimationCts = new();
+		private Rectangle? _activeIndicator;
+
+		private bool _isDisposed;
+		private bool _previousAnimationOngoing;
+
+		// This will be used in the future for horizontal selection indicators.
+		public Orientation NavItemsOrientation { get; set; } = Orientation.Vertical;
+
+		public void Dispose()
+		{
+			_isDisposed = true;
+			CancelPriorAnimation();
+		}
+
+		private void CancelPriorAnimation()
+		{
+			_currentAnimationCts.Cancel();
+			_currentAnimationCts.Dispose();
+			_currentAnimationCts = new CancellationTokenSource();
+		}
+
+		private static Matrix GetOffsetFrom(IVisual ancestor, IVisual visual)
+		{
+			var identity = Matrix.Identity;
+			while (visual != ancestor)
+			{
+				var bounds = visual.Bounds;
+				var topLeft = bounds.TopLeft;
+
+				if (topLeft != new Point())
+				{
+					identity *= Matrix.CreateTranslation(topLeft);
+				}
+
+				if (visual.VisualParent is null)
+				{
+					return Matrix.Identity;
+				}
+
+				visual = visual.VisualParent;
+			}
+
+			return identity;
+		}
+
+		public async void AnimateIndicatorAsync(Rectangle next)
+		{
+			if (_isDisposed)
+			{
+				return;
+			}
+
+			var prevIndicator = _activeIndicator;
+			var nextIndicator = next;
+
+			// user clicked twice
+			if (prevIndicator is null || prevIndicator.Equals(nextIndicator))
+			{
+				return;
+			}
+
+			// Get the common ancestor as a reference point.
+			var commonAncestor = prevIndicator.FindCommonVisualAncestor(nextIndicator);
+
+			// likely being dragged
+			if (commonAncestor is null)
```

Comment:
This shouldn't be null because `FindCommonVisualAncestor` returns non null `IVisual`.
I tested this and it doesn't return null. Hopefully it is not like https://github.com/zkSNACKs/WalletWasabi/pull/6696#discussion_r756955745

Reply:
it's a safeguard, sometimes the control can be null when the control is being dragged. I havent seen it in the wild but it's better to handle it than not at all. Nullable is not infallible and let's not remove nullchecks just because the compiler is complaining about it.

Reply:
@yahiheb the api appears to guarentee it wont return null...

but 

https://github.com/AvaloniaUI/Avalonia/blob/7842883961d094e08e9def7f30cf32fd573179c7/src/Avalonia.Visuals/VisualTree/VisualExtensions.cs#L69

this file doesnt have nullable enabled and so it can return null.